File: nexus/core/rate_limiter.py
"""
Async Semaphore & Token Bucket Rate Limiter
"""

import asyncio
import time
from typing import Dict


class RateLimiter:
    def __init__(self, max_concurrent: int = 3, tokens_per_minute: int = 15):
        self._max_concurrent = max_concurrent
        self._semaphore = asyncio.Semaphore(max_concurrent)
        self._tokens_per_minute = tokens_per_minute
        self._max_tokens = tokens_per_minute + 5
        self._tokens = float(self._max_tokens)
        self._refill_rate = tokens_per_minute / 60.0
        self._last_refill = time.time()
        self._lock = asyncio.Lock()

    async def acquire(self, priority: int = 2):
        async with self._lock:
            self._refill()
            if self._tokens < 1.0:
                wait_time = (1.0 - self._tokens) / self._refill_rate
                await asyncio.sleep(max(0.1, wait_time))
                self._refill()

            self._tokens -= 1.0

        await self._semaphore.acquire()

    def release(self):
        self._semaphore.release()

    def _refill(self):
        now = time.time()
        elapsed = now - self._last_refill
        self._tokens = min(self._max_tokens, self._tokens + (elapsed * self._refill_rate))
        self._last_refill = now


rate_limiter = RateLimiter()

File: js/core/RateLimiter.js
// ============================================================
// NexusResearch -- RateLimiter
// Semaphore + token-bucket hybrid with priority queuing
// ============================================================

/**
 * Controls concurrent API request volume.
 * - Semaphore: limits max concurrent requests
 * - Token bucket: limits requests per time window
 * - Priority queue: higher-priority requests jump the line
 *
 * Priority levels: 1 (highest / command tier) to 3 (lowest / analysis tier)
 */

export class RateLimiter {
  /**
   * @param {Object} options
   * @param {number} [options.maxConcurrent=3]      - Max simultaneous requests
   * @param {number} [options.tokensPerMinute=15]    - Max requests per minute
   * @param {number} [options.burstAllowance=5]      - Extra burst tokens
   */
  constructor({ maxConcurrent = 3, tokensPerMinute = 15, burstAllowance = 5 } = {}) {
    this._maxConcurrent = maxConcurrent;
    this._activeCount = 0;

    // Token bucket
    this._maxTokens = tokensPerMinute + burstAllowance;
    this._tokens = this._maxTokens;
    this._refillRate = tokensPerMinute / 60000; // tokens per ms
    this._lastRefill = Date.now();

    // Priority queue: array of { priority, resolve, reject }
    this._queue = [];
  }

  /**
   * Acquire a slot. Returns a promise that resolves when a slot is available.
   * The returned object has a `release` function that MUST be called when done.
   * @param {number} [priority=2] - 1 (high), 2 (medium), 3 (low)
   * @returns {Promise<{release: Function}>}
   */
  acquire(priority = 2) {
    return new Promise((resolve, reject) => {
      const request = { priority, resolve, reject };
      this._queue.push(request);
      // Sort by priority (lower number = higher priority)
      this._queue.sort((a, b) => a.priority - b.priority);
      this._tryProcess();
    });
  }

  /**
   * Try to process queued requests.
   */
  _tryProcess() {
    this._refillTokens();

    while (this._queue.length > 0 && this._activeCount < this._maxConcurrent && this._tokens >= 1) {
      const request = this._queue.shift();
      this._activeCount++;
      this._tokens--;

      const release = () => {
        this._activeCount = Math.max(0, this._activeCount - 1);
        // Schedule next processing on next microtask
        queueMicrotask(() => this._tryProcess());
      };

      request.resolve({ release });
    }

    // If we have queued items but no tokens, schedule retry when tokens refill
    if (this._queue.length > 0 && this._tokens < 1) {
      const msUntilToken = Math.ceil(1 / this._refillRate);
      setTimeout(() => this._tryProcess(), msUntilToken);
    }
  }

  /**
   * Refill tokens based on elapsed time.
   */
  _refillTokens() {
    const now = Date.now();
    const elapsed = now - this._lastRefill;
    const newTokens = elapsed * this._refillRate;
    this._tokens = Math.min(this._maxTokens, this._tokens + newTokens);
    this._lastRefill = now;
  }

  /**
   * Cancel all queued requests (e.g., on abort).
   */
  cancelAll() {
    const queued = this._queue.splice(0);
    for (const req of queued) {
      req.reject(new Error('Rate limiter cancelled'));
    }
  }

  /** Current stats */
  get stats() {
    return {
      active: this._activeCount,
      queued: this._queue.length,
      tokens: Math.floor(this._tokens),
      maxConcurrent: this._maxConcurrent,
    };
  }
}
